### DISCUSSION TITLE
Allow custom storage adapters

### DISCUSSION THREAD

**Discussion by benhaynes** [directus/directus#3107] [Feature Requests]
To handle additional asset hosting: Cloudinary, Google Drive, etc

Estimated at 2d+

**Comment by rashidpathiyil** [comment]
Is it possible to connect Google Drive to Directus 9?

**Reply by benhaynes** [reply]
Not as of now (in Core), you would need to extend the storage adapters.

**Comment by lazaroofarrill** [comment]
Any news about the possibility of integrating with custom storage adapters?

**Reply by rijkvanzanten** [reply]
@lazaroofarrill Out of curiosity, what storage adapter do you miss currently?

**Reply by lazaroofarrill** [reply]
I'd like to use cloudinary.

**Comment by timstapl** [comment]
@benhaynes  @rijkvanzanten Are you guys open to PRs adding Cloudinary as a storage option?

**Reply by rijkvanzanten** [reply]
Yeah! Though I do have a couple points:

* We should figure out how to sync folders, if we want to go that route
* We should add some configuration to store the generated thumbnails from Directus into a different folder in Cloudinary, as you wouldn't want to have those show up with your regular files
* We need to have https://github.com/directus/directus/pull/8065 (thanks for that!)
* We should allow for a "keep original filename" mode on file upload
* How would we sync changes made in Cloudinary directly back to Directus? (eg when a file name is changed)

Most of these points are relevant for all storage adapters, but seeing cloudinary is an actual external service people would interface with, it becomes a little more important that files are handled neater on the file storage side of things 🙂

**Reply by abdonrd** [reply]
We too would like to be able to use Cloudinary! Thanks guys!

**Comment by AntoineBarroux** [comment]
Hi ! Could be nice to be able to provide whatever you want as a storage adapter, could be useful when you wanna proxy your filesystem with a media REST API for example ! Just something that has some endpoints available to download/upload/delete files, I don't know if I'm understandable. 
But I think it would need to refactor the entire storage provider part so could be expensive

**Comment by crubier** [comment]
I'd like to use [Supabase Storage](https://supabase.io/docs/guides/storage) i'd be willing to code the adapter myself, but how can I then publish it ? From what I understand it is not possible as of today to use third party storage adapters

**Reply by rijkvanzanten** [reply]
Isn't supabase storage just an interface to S3? 🤔 

> From what I understand it is not possible as of today to use third party storage adapters

This is correct, the storage adapters are currently "semi-hardcoded" (eg the storage adapters themselves are packages that can be easily adapted to other services, but they're hardcoded here: https://github.com/directus/directus/blob/bfe83d444e16a6cbc9a392092dc9a2e9f139b917/api/src/storage.ts#L59-L70)

**Reply by ErickLuis00** [reply]
It would be nice to make Storage Adapters an extension type.

https://github.com/directus/directus/discussions/3947

**Reply by akiarostami** [reply]
@rijkvanzanten Even though Supabase stores files on S3, "Supabase storage" is not equal to using S3. For example, they just announced image manipulation functionality on demand for their storage, which only happens if files are uploaded using their interface. 
Any idea if Supabase storage would be available (and when)?

**Reply by rijkvanzanten** [reply]
@akiarostami Hard to give any exact deadlines, but #16825 laid the ground work for more easily creating new native storage adapters, and also provided the foundations to make storage adapters an extension type, so hopefully soon!

**Reply by rijkvanzanten** [reply]
@akiarostami Tracking native supabase storage support in https://github.com/directus/directus/issues/16903 🙂

**Reply by akiarostami** [reply]
Wonderful!

**Comment by rijkvanzanten** [comment]
This is ready to be implemented. The implementation is tracked under https://github.com/directus/directus/issues/16902

**Comment by ghost** [comment]
Hi, i'll like to use directus with nas, possible ??

**Comment by customautosys** [comment]
Hi I understand there are several storage adapters already. However, is it possible for me to write my own custom storage adapter? The reason is that I want to store BLOBs in the database (migrating an existing database which does that). I can extend the existing storage interface and follow the approach of the current storage adapters.

**Reply by rijkvanzanten** [reply]
You can't currently create new custom storage adapters yet
